ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/SemanticTests.scala

```diff
@@ -3,52 +3,340 @@
+        _ <- fetchPaintAgree(ledger, painter, agreement)
+        _ <- fetchPaintAgree(ledger, houseOwner, agreement)
+
+        // The painter sees its new IOU but the house owner cannot see it
+        _ <- fetchIou(ledger, painter, newIou)
+        _ <- fetchIou(ledger, houseOwner, newIou).failed
+
+      } yield {
+        // Nothing to do, all checks have been done in the for-comprehension
+      }
+    }
+
+  private def fetchIou(
+      ledger: LedgerTestContext,
+      party: Primitive.Party,
+      iou: Primitive.ContractId[Iou]): Future[Unit] =
+    for {
+      fetch <- ledger.create(party, FetchIou(party, iou))
+      _ <- ledger.exercise(party, fetch.exerciseFetchIou_Fetch)
+    } yield ()
+
+  private def fetchPaintOffer(
+      ledger: LedgerTestContext,
+      party: Primitive.Party,
+      paintOffer: Primitive.ContractId[PaintOffer]
+  ): Future[Unit] =
+    for {
+      fetch <- ledger.create(party, FetchPaintOffer(party, paintOffer))
+      _ <- ledger.exercise(party, fetch.exerciseFetchPaintOffer_Fetch)
+    } yield ()
+
+  private def fetchPaintAgree(
+      ledger: LedgerTestContext,
+      party: Primitive.Party,
+      agreement: Primitive.ContractId[PaintAgree]
+  ): Future[Unit] =
+    for {
+      fetch <- ledger.create(party, FetchPaintAgree(party, agreement))
+      _ <- ledger.exercise(party, fetch.exerciseFetchPaintAgree_Fetch)
+    } yield ()
+
+  /*
+   * Divulgence
+   */
+
+  private[this] val divulgence =
+    LedgerTest("SemanticDivulgence", "Respect divulgence rules") { ledger =>
+      for {
+        Vector(issuer, owner, delegate) <- ledger.allocateParties(3)
+        token <- ledger.create(issuer, Token(issuer, owner, 1))
+        delegation <- ledger.create(owner, Delegation(owner, delegate))
+
+        // The owner tries to delegate with a non-consuming choice, which actually doesn't
+        noDivulgeToken <- ledger.create(owner, Delegation(owner, delegate))
+        _ <- ledger.exercise(owner, noDivulgeToken.exerciseDelegation_Wrong_Divulge_Token(_, token))
+        _ <- ledger.exercise(delegate, delegation.exerciseDelegation_Token_Consume(_, token)).failed
+
+        // Successful divulgence and delegation
+        divulgeToken <- ledger.create(owner, Delegation(owner, delegate))
+        _ <- ledger.exercise(owner, divulgeToken.exerciseDelegation_Divulge_Token(_, token))
+        _ <- ledger.exercise(delegate, delegation.exerciseDelegation_Token_Consume(_, token))
+      } yield {}
+    }
+
+  /*
+   * Contract keys
+   */
+
+  private[this] val contractKeys =
+    LedgerTest("SemanticContractKeys", "Perform correctly operations based on contract keys") {
+      ledger =>
+        for {
+          Vector(bank, accountHolder) <- ledger.allocateParties(2)
+          accountTemplate = Account(bank, accountHolder, DamlTuple2("CH", 123))
+          accountKey = DamlTuple2(bank, "CH123") // key (bank, accountNumber._1 <> show (this.accountNumber._2)) : (Party, Text)
```

Comment:
@stefanobaghino-da : what does the comment here mean?

Reply:
It's the way in which the key is computed on DAML. I left it there for reference. Unfortunately I don't think there is a way to do it organically without replicating the logic to get the key. Not that readable though, thanks for the comment, I'll make it explicit.

Reply:
Addressed by https://github.com/digital-asset/daml/pull/2718/commits/7a623c3d5de46231da70f08dc86df5621b56bdbc